From efbbf22829a894cda8e9e6f4a49004f06d3bf483 Mon Sep 17 00:00:00 2001
From: Deepti Jaggi <quic_djaggi@quicinc.com>
Date: Tue, 22 Oct 2024 16:34:46 -0700
Subject: [PATCH] PENDING: cpufreq: scmi: Skip SCMI devices that aren't used by
 the CPUs

We may have multiple SCMI devices that expose PERF protocols. However,
only one of them can be used to control the CPUs. Return early from
probes called for SCMI devices that aren't used by the CPUs.

Change-Id: Ib0a181fb2825f2beb74a3c697c4437790e6ca707
Signed-off-by: Mike Tipton <quic_mdtipton@quicinc.com>
Signed-off-by: Deepti Jaggi <quic_djaggi@quicinc.com>
---
 drivers/cpufreq/scmi-cpufreq.c | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/drivers/cpufreq/scmi-cpufreq.c b/drivers/cpufreq/scmi-cpufreq.c
index 5c0f4e2c64c4..5e088dc0722a 100644
--- a/drivers/cpufreq/scmi-cpufreq.c
+++ b/drivers/cpufreq/scmi-cpufreq.c
@@ -317,6 +317,31 @@ static struct cpufreq_driver scmi_cpufreq_driver = {
 	.register_em	= scmi_cpufreq_register_em,
 };
 
+static bool scmi_dev_used_by_cpus(struct device *scmi_dev)
+{
+	struct of_phandle_args args;
+	struct device_node *np;
+	struct device *cpu_dev;
+	int cpu, ret;
+
+	for_each_possible_cpu(cpu) {
+		cpu_dev = get_cpu_device(cpu);
+		if (!cpu_dev)
+			continue;
+
+		np = cpu_dev->of_node;
+
+		ret = of_parse_phandle_with_args(np, "clocks", "#clock-cells", 0, &args);
+		if (ret < 0)
+			continue;
+
+		if (args.np == scmi_dev->of_node)
+			return true;
+	}
+
+	return false;
+}
+
 static int scmi_cpufreq_probe(struct scmi_device *sdev)
 {
 	int ret;
@@ -328,6 +353,9 @@ static int scmi_cpufreq_probe(struct scmi_device *sdev)
 	if (!handle)
 		return -ENODEV;
 
+	if (!scmi_dev_used_by_cpus(dev))
+		return 0;
+
 	perf_ops = handle->devm_protocol_get(sdev, SCMI_PROTOCOL_PERF, &ph);
 	if (IS_ERR(perf_ops))
 		return PTR_ERR(perf_ops);
-- 
2.25.1

